File: python/spider_browser/protocol/transport.py
"""WebSocket transport to Spider's browser fleet."""

from __future__ import annotations

import asyncio
import json
from typing import Any, Callable, Optional
from urllib.parse import urlencode

import websockets
import websockets.client

from ..events.emitter import SpiderEventEmitter
from ..events.types import BrowserType
from ..utils.errors import ConnectionError, AuthError, BackendUnavailableError, RateLimitError, TimeoutError
from ..utils.logger import logger


class Transport:
    """
    WebSocket transport to Spider's browser fleet.

    Connects to ``wss://browser.spider.cloud/v1/browser?token=TOKEN&browser=TYPE``,
    handles reconnection on browser switch, and dispatches raw messages.
    """

    def __init__(
        self,
        *,
        api_key: str,
        server_url: str,
        browser: BrowserType,
        url: Optional[str] = None,
        captcha: str = "solve",
        stealth: int = 0,
        record: bool = False,
        mode: Optional[str] = None,
        hedge: bool = False,
        country: Optional[str] = None,
        proxy_url: Optional[str] = None,
        connect_timeout_ms: int = 30_000,
        command_timeout_ms: int = 30_000,
        emitter: SpiderEventEmitter,
    ) -> None:
        self._api_key = api_key
        self._server_url = server_url.rstrip("/")
        self._current_browser: BrowserType = "chrome-h" if browser == "auto" else browser
        self._url = url
        self._captcha = captcha
        self._stealth_level = stealth
        self._record = record
        self._mode = mode
        self._hedge = hedge
        self._country = country
        self._proxy_url = proxy_url
        self._connect_timeout_ms = connect_timeout_ms
        self._command_timeout_ms = command_timeout_ms
        self._emitter = emitter
        self._ws: Optional[websockets.client.WebSocketClientProtocol] = None
        self._message_handler: Optional[Callable[[str], None]] = None
        self._recv_task: Optional[asyncio.Task[None]] = None
        self._generation = 0

        # Upgrade response headers
        self._upgrade_credits: Optional[float] = None
        self._upgrade_stealth_tier: Optional[int] = None
        self._upgrade_proxy_tier: Optional[int] = None
        # Session metering
        self._session_credits_used: Optional[float] = None

    @property
    def browser(self) -> BrowserType:
        return self._current_browser

    @property
    def connected(self) -> bool:
        return self._ws is not None and self._ws.open

    @property
    def stealth_level(self) -> int:
        return self._stealth_level

    @stealth_level.setter
    def stealth_level(self, level: int) -> None:
        self._stealth_level = max(0, min(3, level))

    @property
    def upgrade_credits(self) -> Optional[float]:
        """Credits remaining from the WebSocket upgrade response."""
        return self._upgrade_credits

    @property
    def upgrade_stealth_tier(self) -> Optional[int]:
        """Active stealth tier from the WebSocket upgrade response."""
        return self._upgrade_stealth_tier

    @property
    def upgrade_proxy_tier(self) -> Optional[int]:
        """Proxy tier bucket from the WebSocket upgrade response (1=cheap, 2=mid, 3=strong)."""
        return self._upgrade_proxy_tier

    @property
    def session_credits_used(self) -> Optional[float]:
        """Credits consumed during this session (from server Spider.metering event)."""
        return self._session_credits_used

    @property
    def command_timeout_ms(self) -> int:
        return self._command_timeout_ms

    async def request_metering(self, timeout_ms: int = 3000) -> float:
        """Request the current session cost from the server via Spider.getMetering."""
        if not self._ws or not self._ws.open:
            return self._session_credits_used or 0

        metering_id = 2147483640  # High ID to avoid collisions

        fut: asyncio.Future[float] = asyncio.get_event_loop().create_future()
        orig_handler = self._message_handler

        def _intercept(data: str) -> None:
            if f'"id":{metering_id}' in data:
                try:
                    msg = json.loads(data)
                    if msg.get("id") == metering_id and msg.get("result", {}).get("credits_used") is not None:
                        self._session_credits_used = msg["result"]["credits_used"]
                        self._message_handler = orig_handler
                        if not fut.done():
                            fut.set_result(msg["result"]["credits_used"])
                        return
                except Exception:
                    pass
            if orig_handler:
                orig_handler(data)

        self._message_handler = _intercept

        try:
            await self.send_async(json.dumps({"id": metering_id, "method": "Spider.getMetering"}))
            return await asyncio.wait_for(fut, timeout=timeout_ms / 1000)
        except (asyncio.TimeoutError, Exception):
            self._message_handler = orig_handler
            return self._session_credits_used or 0

    def on_message(self, handler: Callable[[str], None]) -> None:
        self._message_handler = handler

    async def connect(self, max_attempts: int = 3) -> None:
        """Connect to the WebSocket endpoint with retry."""
        if self._ws and self._ws.open:
            return

        last_error: Optional[Exception] = None
        for attempt in range(1, max_attempts + 1):
            try:
                await self._connect_internal()
                return
            except AuthError:
                raise
            except Exception as exc:
                last_error = exc
                if attempt < max_attempts:
                    backoff = 0.5 * attempt
                    logger.warning(
                        f"connect attempt {attempt}/{max_attempts} failed, retrying in {backoff}s: {exc}"
                    )
                    await asyncio.sleep(backoff)
        raise last_error  # type: ignore

    async def reconnect(self, browser: BrowserType) -> None:
        prev = self._current_browser
        self._current_browser = browser
        await self.close()
        logger.info(f"switching browser: {prev} -> {browser}")
        await self._connect_internal()

    def send(self, data: str) -> None:
        if not self._ws or not self._ws.open:
            raise ConnectionError("WebSocket is not connected")
        asyncio.get_event_loop().create_task(self._ws.send(data))

    async def send_async(self, data: str) -> None:
        if not self._ws or not self._ws.open:
            raise ConnectionError("WebSocket is not connected")
        await self._ws.send(data)

    async def close(self) -> None:
        if self._recv_task:
            self._recv_task.cancel()
            try:
                await self._recv_task
            except (asyncio.CancelledError, Exception):
                pass
            self._recv_task = None
        if self._ws:
            try:
                await self._ws.close()
            except Exception:
                pass
            self._ws = None

    def _build_url(self) -> str:
        params: dict[str, str] = {"token": self._api_key}
        if self._current_browser != "auto":
            params["browser"] = self._current_browser
        if self._url:
            params["url"] = self._url
        if self._captcha and self._captcha != "off":
            params["ai_captcha"] = self._captcha
        if self._stealth_level > 0:
            params["s"] = str(self._stealth_level)
        if self._hedge:
            params["hedge"] = "true"
        if self._record:
            params["record"] = "true"
        if self._mode:
            params["mode"] = self._mode
        if self._country:
            params["country"] = self._country
        if self._proxy_url:
            params["proxy_url"] = self._proxy_url
        return f"{self._server_url}/v1/browser?{urlencode(params)}"

    async def _connect_internal(self) -> None:
        self._generation += 1
        gen = self._generation

        url = self._build_url()
        safe_url = url.split("token=")[0] + "token=***"
        logger.debug(f"connecting to {safe_url}")

        timeout_s = self._connect_timeout_ms / 1000

        try:
            self._ws = await asyncio.wait_for(
                websockets.connect(url, max_size=50 * 1024 * 1024, open_timeout=timeout_s),
                timeout=timeout_s,
            )
        except asyncio.TimeoutError:
            raise TimeoutError(f"WebSocket connection timeout ({self._connect_timeout_ms}ms)")
        except Exception as exc:
            msg = str(exc)
            if "429" in msg:
                raise RateLimitError("Server at capacity (429)")
            raise ConnectionError(f"WebSocket error: {exc}")

        self._emitter.emit("ws.open")
        logger.info(f"connected (browser={self._current_browser}, stealth={self._stealth_level})")
        self._recv_task = asyncio.get_event_loop().create_task(self._recv_loop(gen))

    async def _recv_loop(self, gen: int) -> None:
        assert self._ws is not None
        try:
            async for raw in self._ws:
                # Guard: ignore messages from stale WebSocket instances
                if gen != self._generation:
                    return
                msg = raw if isinstance(raw, str) else raw.decode("utf-8", errors="replace")

                # Intercept Spider.* events from the server
                if '"Spider.' in msg:
                    if self._handle_spider_transport_event(msg):
                        continue

                if self._message_handler:
                    self._message_handler(msg)
        except websockets.exceptions.ConnectionClosedError as exc:
            if gen == self._generation:
                code = exc.code
                reason = str(exc.reason)
                self._emitter.emit("ws.close", {"code": code, "reason": reason})
        except asyncio.CancelledError:
            pass
        except Exception as exc:
            if gen == self._generation:
                self._emitter.emit("ws.error", {"error": str(exc)})

    def _handle_spider_transport_event(self, data: str) -> bool:
        """Handle Spider.* events at the transport level. Returns True if consumed."""
        try:
            msg = json.loads(data)
            method = msg.get("method")
            params = msg.get("params", {})

            if method == "Spider.screencastFrame":
                self._emitter.emit("screencast.frame", params)
                return True
            if method == "Spider.interactionEvents":
                self._emitter.emit("screencast.interactionEvents", params)
                return True
            if method == "Spider.rrwebEvents":
                self._emitter.emit("screencast.rrwebEvents", params)
                return True
            if method == "Spider.recordingStarted":
                self._emitter.emit("recording.started", params)
                return True
            if method == "Spider.recordingCompleted":
                self._emitter.emit("recording.completed", params)
                return True
            if method == "Spider.metering":
                credits_used = params.get("credits_used")
                if credits_used is not None:
                    self._session_credits_used = credits_used
                    self._emitter.emit("metering", {
                        "credits": self._upgrade_credits or 0,
                        "rate": self._upgrade_stealth_tier or 0,
                        "session_credits_used": credits_used,
                    })
                    return True
        except Exception:
            pass
        return False

File: go/transport.go
package spiderbrowser

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"github.com/gorilla/websocket"
)

// TransportOptions configures the WebSocket connection to Spider.
type TransportOptions struct {
	APIKey          string
	ServerURL       string
	Browser         string
	URL             string
	Captcha         string // "off", "detect", or "solve"
	StealthLevel    int
	Country         string
	ProxyURL        string
	ConnectTimeout  time.Duration
	CommandTimeout  time.Duration
	Hedge           bool
	Record          bool
	Mode            string // "scraping" or "cua"
}

func (o *TransportOptions) defaults() {
	if o.ServerURL == "" {
		o.ServerURL = "wss://browser.spider.cloud"
	}
	if o.Browser == "" {
		o.Browser = "auto"
	}
	if o.Captcha == "" {
		o.Captcha = "solve"
	}
	if o.ConnectTimeout == 0 {
		o.ConnectTimeout = 30 * time.Second
	}
	if o.CommandTimeout == 0 {
		o.CommandTimeout = 30 * time.Second
	}
}

// Transport manages the WebSocket connection to Spider's browser fleet.
type Transport struct {
	mu             sync.Mutex
	conn           *websocket.Conn
	opts           TransportOptions
	emitter        *EventEmitter
	currentBrowser string
	stealthLevel   int
	generation     int64
	messageHandler func(data string)

	upgradeCredits    *float64
	upgradeStealthTier *int
	upgradeProxyTier  *int
	sessionCreditsUsed *float64

	closed int32 // atomic
}

// NewTransport creates a new WebSocket transport.
func NewTransport(opts TransportOptions, emitter *EventEmitter) *Transport {
	opts.defaults()
	browser := opts.Browser
	if browser == "auto" {
		browser = "chrome-h"
	}
	return &Transport{
		opts:           opts,
		emitter:        emitter,
		currentBrowser: browser,
		stealthLevel:   opts.StealthLevel,
	}
}

// Browser returns the current browser type.
func (t *Transport) Browser() string {
	t.mu.Lock()
	defer t.mu.Unlock()
	return t.currentBrowser
}

// Connected returns whether the WebSocket is currently open.
func (t *Transport) Connected() bool {
	t.mu.Lock()
	defer t.mu.Unlock()
	return t.conn != nil
}

// StealthLevel returns the active stealth level.
func (t *Transport) StealthLevel() int {
	t.mu.Lock()
	defer t.mu.Unlock()
	return t.stealthLevel
}

// SetStealthLevel sets the stealth level (clamped to 0-3).
func (t *Transport) SetStealthLevel(level int) {
	t.mu.Lock()
	defer t.mu.Unlock()
	if level < 0 {
		level = 0
	}
	if level > 3 {
		level = 3
	}
	t.stealthLevel = level
}

// UpgradeCredits returns credits remaining from the last upgrade response.
func (t *Transport) UpgradeCredits() *float64 { return t.upgradeCredits }

// SessionCreditsUsed returns credits consumed during this session.
func (t *Transport) SessionCreditsUsed() *float64 { return t.sessionCreditsUsed }

// OnMessage sets the handler that receives raw JSON messages from the WebSocket.
func (t *Transport) OnMessage(handler func(data string)) {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.messageHandler = handler
}

// Connect establishes the WebSocket connection with retry.
func (t *Transport) Connect(maxAttempts int) error {
	if maxAttempts <= 0 {
		maxAttempts = 3
	}
	var lastErr error
	for attempt := 1; attempt <= maxAttempts; attempt++ {
		err := t.connectInternal()
		if err == nil {
			return nil
		}
		lastErr = err
		if _, ok := err.(*AuthError); ok {
			return err
		}
		if attempt < maxAttempts {
			backoff := time.Duration(500*attempt) * time.Millisecond
			defaultLogger.warn(fmt.Sprintf("connect attempt %d/%d failed, retrying in %v", attempt, maxAttempts, backoff), "error", lastErr)
			time.Sleep(backoff)
		}
	}
	return lastErr
}

// Reconnect closes and reconnects with a different browser type.
func (t *Transport) Reconnect(browser string) error {
	t.mu.Lock()
	prev := t.currentBrowser
	t.currentBrowser = browser
	t.mu.Unlock()

	t.Close()
	defaultLogger.info(fmt.Sprintf("switching browser: %s -> %s", prev, browser))
	return t.connectInternal()
}

// Send sends a raw JSON string through the WebSocket.
func (t *Transport) Send(data string) error {
	t.mu.Lock()
	conn := t.conn
	t.mu.Unlock()

	if conn == nil {
		return newConnectionError("WebSocket is not connected")
	}
	return conn.WriteMessage(websocket.TextMessage, []byte(data))
}

// Close closes the WebSocket connection.
func (t *Transport) Close() {
	if !atomic.CompareAndSwapInt32(&t.closed, 0, 1) {
		// Allow re-closing after reconnect
	}
	t.mu.Lock()
	conn := t.conn
	t.conn = nil
	t.mu.Unlock()

	if conn != nil {
		conn.Close()
	}
	atomic.StoreInt32(&t.closed, 0) // Reset for potential reconnect
}

// RequestMetering sends Spider.getMetering and waits for the response.
func (t *Transport) RequestMetering(timeout time.Duration) (float64, error) {
	if timeout == 0 {
		timeout = 3 * time.Second
	}
	t.mu.Lock()
	conn := t.conn
	t.mu.Unlock()
	if conn == nil {
		if t.sessionCreditsUsed != nil {
			return *t.sessionCreditsUsed, nil
		}
		return 0, nil
	}

	meteringID := 2147483640
	cmd, _ := json.Marshal(map[string]any{
		"id":     meteringID,
		"method": "Spider.getMetering",
	})

	ch := make(chan float64, 1)

	// Temporarily intercept in the message handler
	origHandler := t.messageHandler
	t.mu.Lock()
	t.messageHandler = func(data string) {
		if strings.Contains(data, fmt.Sprintf(`"id":%d`, meteringID)) {
			var msg struct {
				ID     int `json:"id"`
				Result struct {
					CreditsUsed float64 `json:"credits_used"`
				} `json:"result"`
			}
			if json.Unmarshal([]byte(data), &msg) == nil && msg.ID == meteringID {
				cu := msg.Result.CreditsUsed
				t.sessionCreditsUsed = &cu
				t.mu.Lock()
				t.messageHandler = origHandler
				t.mu.Unlock()
				select {
				case ch <- cu:
				default:
				}
				return
			}
		}
		if origHandler != nil {
			origHandler(data)
		}
	}
	t.mu.Unlock()

	if err := t.Send(string(cmd)); err != nil {
		t.mu.Lock()
		t.messageHandler = origHandler
		t.mu.Unlock()
		if t.sessionCreditsUsed != nil {
			return *t.sessionCreditsUsed, nil
		}
		return 0, nil
	}

	select {
	case credits := <-ch:
		return credits, nil
	case <-time.After(timeout):
		t.mu.Lock()
		t.messageHandler = origHandler
		t.mu.Unlock()
		if t.sessionCreditsUsed != nil {
			return *t.sessionCreditsUsed, nil
		}
		return 0, nil
	}
}

func (t *Transport) buildURL() string {
	base := strings.TrimRight(t.opts.ServerURL, "/")
	params := url.Values{}
	params.Set("token", t.opts.APIKey)

	t.mu.Lock()
	browser := t.currentBrowser
	stealth := t.stealthLevel
	t.mu.Unlock()

	if browser != "auto" {
		params.Set("browser", browser)
	}
	if t.opts.URL != "" {
		params.Set("url", t.opts.URL)
	}
	if t.opts.Captcha != "" && t.opts.Captcha != "off" {
		params.Set("ai_captcha", t.opts.Captcha)
	}
	if stealth > 0 {
		params.Set("s", strconv.Itoa(stealth))
	}
	if t.opts.Hedge {
		params.Set("hedge", "true")
	}
	if t.opts.Record {
		params.Set("record", "true")
	}
	if t.opts.Mode != "" {
		params.Set("mode", t.opts.Mode)
	}
	if t.opts.Country != "" {
		params.Set("country", t.opts.Country)
	}
	if t.opts.ProxyURL != "" {
		params.Set("proxy_url", t.opts.ProxyURL)
	}
	return fmt.Sprintf("%s/v1/browser?%s", base, params.Encode())
}

func (t *Transport) connectInternal() error {
	gen := atomic.AddInt64(&t.generation, 1)

	wsURL := t.buildURL()
	maskedURL := strings.Replace(wsURL, t.opts.APIKey, "***", 1)
	defaultLogger.debug(fmt.Sprintf("connecting to %s", maskedURL))

	dialer := websocket.Dialer{
		HandshakeTimeout: t.opts.ConnectTimeout,
	}

	conn, resp, err := dialer.Dial(wsURL, nil)
	if err != nil {
		if resp != nil {
			switch resp.StatusCode {
			case 401, 402:
				return newAuthError(fmt.Sprintf("Authentication failed (%d)", resp.StatusCode))
			case 429:
				return newRateLimitError("Server at capacity (429)")
			case 503:
				return newBackendUnavailableError("Backend unavailable (503)")
			}
		}
		return newConnectionError(fmt.Sprintf("WebSocket connection failed: %v", err))
	}

	// Parse upgrade headers
	if resp != nil {
		if sc := resp.Header.Get("X-Sc"); sc != "" {
			if v, err := strconv.ParseFloat(sc, 64); err == nil {
				t.upgradeCredits = &v
			}
		}
		if sr := resp.Header.Get("X-Sr"); sr != "" {
			if v, err := strconv.Atoi(sr); err == nil {
				t.upgradeStealthTier = &v
			}
		}
		if pt := resp.Header.Get("X-Pt"); pt != "" {
			if v, err := strconv.Atoi(pt); err == nil {
				t.upgradeProxyTier = &v
			}
		}
		if t.upgradeCredits != nil {
			rate := 0
			if t.upgradeStealthTier != nil {
				rate = *t.upgradeStealthTier
			}
			t.emitter.Emit("metering", map[string]any{
				"credits": *t.upgradeCredits,
				"rate":    rate,
			})
		}
	}

	t.mu.Lock()
	t.conn = conn
	t.mu.Unlock()

	t.emitter.Emit("ws.open", map[string]any{})
	defaultLogger.info(fmt.Sprintf("connected (browser=%s, stealth=%d)", t.currentBrowser, t.stealthLevel))

	// Start read loop in goroutine
	go t.readLoop(conn, gen)

	return nil
}

func (t *Transport) readLoop(conn *websocket.Conn, gen int64) {
	defer func() {
		t.mu.Lock()
		if t.conn == conn {
			t.conn = nil
		}
		t.mu.Unlock()
	}()

	for {
		_, message, err := conn.ReadMessage()
		if err != nil {
			// Only emit close for current generation
			if atomic.LoadInt64(&t.generation) == gen {
				code := 0
				reason := err.Error()
				if closeErr, ok := err.(*websocket.CloseError); ok {
					code = closeErr.Code
					reason = closeErr.Text
				}
				t.emitter.Emit("ws.close", map[string]any{
					"code":   code,
					"reason": reason,
				})
			}
			return
		}

		// Guard: ignore messages from stale WebSocket instances
		if atomic.LoadInt64(&t.generation) != gen {
			continue
		}

		str := string(message)

		// Intercept Spider.* events
		if strings.Contains(str, `"Spider.`) {
			if t.handleSpiderMessage(str) {
				continue
			}
		}

		// Intercept Spider.metering events
		if strings.Contains(str, `"Spider.metering"`) {
			var msg struct {
				Method string `json:"method"`
				Params struct {
					CreditsUsed float64 `json:"credits_used"`
				} `json:"params"`
			}
			if json.Unmarshal(message, &msg) == nil && msg.Method == "Spider.metering" {
				cu := msg.Params.CreditsUsed
				t.sessionCreditsUsed = &cu
				credits := float64(0)
				if t.upgradeCredits != nil {
					credits = *t.upgradeCredits
				}
				rate := 0
				if t.upgradeStealthTier != nil {
					rate = *t.upgradeStealthTier
				}
				t.emitter.Emit("metering", map[string]any{
					"credits":              credits,
					"rate":                 rate,
					"session_credits_used": cu,
				})
				continue
			}
		}

		t.mu.Lock()
		handler := t.messageHandler
		t.mu.Unlock()
		if handler != nil {
			handler(str)
		}
	}
}

func (t *Transport) handleSpiderMessage(data string) bool {
	var msg struct {
		Method string         `json:"method"`
		Params map[string]any `json:"params"`
	}
	if err := json.Unmarshal([]byte(data), &msg); err != nil {
		return false
	}

	switch msg.Method {
	case "Spider.screencastFrame":
		t.emitter.Emit("screencast.frame", msg.Params)
		return true
	case "Spider.interactionEvents":
		t.emitter.Emit("screencast.interactionEvents", msg.Params)
		return true
	case "Spider.rrwebEvents":
		t.emitter.Emit("screencast.rrwebEvents", msg.Params)
		return true
	case "Spider.recordingStarted":
		t.emitter.Emit("recording.started", msg.Params)
		return true
	case "Spider.recordingCompleted":
		t.emitter.Emit("recording.completed", msg.Params)
		return true
	}
	return false
}

// httpHeaderToMap converts http.Header to a simple map for upgrade header parsing.
func httpHeaderToMap(h http.Header) map[string]string {
	m := make(map[string]string)
	for k, v := range h {
		if len(v) > 0 {
			m[k] = v[0]
		}
	}
	return m
}
